### PR TITLE
feat: add reproduce area options in debug-map menu

### PR DESCRIPTION
**Pull request by ekaratzas**
Extends the map submenu of the debug menu with an option to force monsters in a rectangular area to reproduce.

cow goes moo, sheep goes bah

## Purpose of change

Purpose is to ease debugging of hard-to-reproduce corner-cases such as what happens when a monster reproduces and spawns offspring.

## Describe the solution

Solution works by extending map sub menu of debug menu with an option to force all monsters in an area to reproduce. This can actually be triggered multiple times in the row in order to quickly create a huge flock of creatures out of a specific parent. The forcing of reproduction works by splitting try_reproduce into 2 parts, try() and actual reproduce()

Note that the solution does not cause eggs to hatch as that is based on the rot-timer functionality which is in a different part of the codebase altogether. TODO for future PR.

## Testing

spawn creatures. use new debug command to get them to spawn other creatures. then use debug command to kill in area to check that we didnt break that as well due to refactoring. moo

### CODE REVIEW

**Comment by ekaratzas** [comment]
@chaosvolt / @scarf005  fyi

**Review by scarf005 — CHANGES_REQUESTED** [review]
build fails in lua bindings due to change in function signature.

```
/home/scarf/repo/cata/Cataclysm/src/catalua_bindings_creature.cpp:285:9: error: no matching function for call to 'resolve'
  285 |         SET_FX_T( try_reproduce, void() );
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/scarf/repo/cata/Cataclysm/src/catalua_bindings_utils.h:28:9: note: expanded from macro 'SET_FX_T'
   28 |         sol::resolve< func_type >( &UT_CLASS::func_name))
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~
/home/scarf/repo/cata/Cataclysm/src/sol/sol.hpp:19779:14: note: candidate function template not viable: no known conversion from 'void (monster::*)(const bool)' to 'void (*)()' for 1st argument
 19779 |         inline Sig* resolve(Sig* fun_ptr) {
       |                     ^       ~~~~~~~~~~~~
/home/scarf/repo/cata/Cataclysm/src/sol/sol.hpp:19774:14: note: candidate template ignored: could not match 'R (*)(void (*)(), Args...)' against 'void (monster::*)(const bool)'
 19774 |         inline auto resolve(R fun_ptr(Args...)) -> R (*)(Args...) {
       |                     ^
/home/scarf/repo/cata/Cataclysm/src/sol/sol.hpp:19784:14: note: candidate template ignored: could not match 'void (*)()' against 'bool'
 19784 |         inline auto resolve(R (C::*mem_ptr)(Args...)) -> R (C::*)(Args...) {
       |                     ^
/home/scarf/repo/cata/Cataclysm/src/sol/sol.hpp:19789:17: note: candidate template ignored: failed template argument deduction
 19789 |         inline Sig C::*resolve(Sig C::*mem_ptr) {
       |                        ^
/home/scarf/repo/cata/Cataclysm/src/sol/sol.hpp:19794:14: note: candidate template ignored: substitution failure [with Sig = <void ()>, F = void (monster::*)(const bool)]: no matching function for call to 'resolve_i'
 19794 |         inline auto resolve(F&& f) -> decltype(detail::resolve_i(types<Sig...>(), std::forward<F>(f))) {
       |                     ^                          ~~~~~~
1 error generated.
```

File: src/monster.h [hunk 1]
```diff
@@ -122,7 +122,8 @@ class monster : public Creature, public location_visitable<monster>
         int get_upgrade_time() const;
         void allow_upgrade();
         void try_upgrade( bool pin_time );
-        void try_reproduce();
+        // check if monster should reproduce. boolean parameter can force reproduction
+        void try_reproduce( const bool force_reproduce = false );
```
Comment (scarf005) [inline src/monster.h:126]: i think it'd be better to add `void reproduce()` than to overload it with boolean parameter so one function does one thing.

Comment (scarf005) [reply src/monster.h:126]: opened PR for suggested changes in https://github.com/ekaratzas/Cataclysm-BN/pull/1

**Comment by ekaratzas** [comment]
@scarf005 just tested again on top of latest changes.

**Comment by scarf005** [comment]
image

you can also re-request reviews here.

**Comment by ekaratzas** [comment]
good point, still rusty with github

**Review by scarf005 — COMMENTED** [review]
image

chicken eggs spawned via this way seems to have negative birthday.

**Comment by ekaratzas** [comment]
@scarf005 did you reproduce this on top of just this PR or with extra code you added to force rotting of eggs?

**Comment by scarf005** [comment]
with the extra code. i guess this is because baby timer isn't initialized cotrectly

**Comment by ekaratzas** [comment]
yes just saw that

**Comment by ekaratzas** [comment]
@scarf005 curious, I can't seem to repro your issue. I tried causing friendly and not friendly chickens to spawn eggs and i move to their tile to pick up the eggs 1 turn afterwards and examine the items. The age always seems to be '1', not a negative value. 

What am I missing in the repro steps? :S

egg_age

**Comment by scarf005** [comment]
hmm strange, can't reproduce.

**Review by scarf005 — APPROVED** [review]

**Pull request merged**